File: 13 Fibonacci Numbers/Ladder/ladder.js
function solution(A, B) {
  const aLen = A.length;
  const result = [];
  let max = 0;
  const steps = [1, 1];
  let maxB = 0;

  for (let i = 0; i < aLen; i += 1) {
    max = Math.max(max, A[i]);
    maxB = Math.max(maxB, B[i]);
  }

  let y = 1;
  while ((y += 1) <= max) {
    steps[y] = (steps[y - 1] + steps[y - 2]) % (2 ** maxB);
  }

  for (let x = 0; x < aLen; x += 1) {
    const div = steps[A[x]] & (2 ** B[x] - 1);
    result.push(div);
  }

  return result;
}

solution([4, 4, 5, 5, 1], [3, 2, 4, 3, 1]); // [5, 1, 8, 0, 1],

File: 13 Fibonacci Numbers/Ladder/ladder.java
import java.util.Arrays;

class Solution {
  public int[] solution(int[] A, int[] B) {
    int[] result = new int[A.length];
    int[] steps = new int[A.length + 1];
    steps[0] = 1;
    steps[1] = 1;

    int max = Arrays.stream(A).max().getAsInt();
    int maxB = Arrays.stream(B).max().getAsInt();

    int y = 1;
    while ((y += 1) <= max) {
      steps[y] = (steps[y - 1] + steps[y - 2]) % (int) Math.pow(2, maxB);
    }

    for (int x = 0; x < A.length; x += 1) {
      int div = (int) Math.pow(2, B[x]) - 1;
      result[x] = steps[A[x]] & div;
    }

    return result;
  }
}